xheap.py

```diff
@@ -181,8 +170,12 @@ def _get_indexes(self):
         return indexes
 
     def __setitem__(self, key, value):
-        self._indexes[value] = key
+        if key < 0:
+            key += len(self)
+        if key < 0:
+            raise IndexError
         super(RemovalHeap, self).__setitem__(key, value)
+        self._indexes[value] = key
```

Comment:
only assign `self._indexes[value] = key` after the `super(RemovalHeap, self).__setitem__(key, value)` has completed without any exception.